internal/rest/types/hooks.go

```diff
@@ -0,0 +1,45 @@
+package types
+
+// HookType represents the various types of hooks available to microcluster.
+type HookType string
+
+const (
+	// OnStart is run after the daemon is started.
+	OnStart HookType = "on-start"
+
+	// PreBootstrap is run before the daemon is initialized and bootstrapped.
+	PreBootstrap HookType = "pre-bootstrap"
+
+	// PostBootstrap is run after the daemon is initialized and bootstrapped.
+	PostBootstrap HookType = "post-bootstrap"
+
+	// PreJoin is run after the daemon is initialized and joined the cluster but before existing members triggered
+	// their 'OnNewMember' hooks.
+	PreJoin HookType = "pre-join"
+
+	// PostJoin is run after the daemon is initialized, joined the cluster and existing members triggered
+	// their 'OnNewMember' hooks.
+	PostJoin HookType = "post-join"
+
+	// PreRemove is run on a cluster member just before it is removed from the cluster.
+	PreRemove HookType = "pre-remove"
+
+	// PostRemove is run on all other peers after one is removed from the cluster.
+	PostRemove HookType = "post-remove"
+
+	// OnNewMember is run on each peer after a new cluster member has joined and executed their 'PreJoin' hook.
+	OnNewMember HookType = "on-new-member"
+
+	// OnHeartbeat is run after a successful heartbeat round.
+	OnHeartbeat HookType = "on-heartbeat"
+)
+
+// HookConfig holds configuration pertaining to different hooks.
+type HookConfig struct {
+	// Force represents whether or not the hook should consider validation.
+	Force bool `json:"force" yaml:"force"`
+
+	// Name represents the name of the node that triggered the hook.
+	// This is distinct from the node that the hook is triggered on.
+	Name string `json:"name" yaml:"name"`
```

Comment:
Or perhaps `CallerNodeName`, `CallerMemberName`

Reply:
Issue with an approach like that is that the client function would have to accept `any` as the type for the payload and we would have to validate the payload either in the client function or on the server side, or both.

Reply:
What about returning an interface from the client function? It could have the function `Triggerer() string` so for each specific hook type we can set special implementation on how to retrieve those details from the underlying data structure? And there could be a `Type() string` which returns the name of the hook so that it can be casted into any of the underlying structs if necessary?

Reply:
At that point we might as well just have different client functions for each hook that we want, rather than directly specifying the hooktype to the client function.